WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSource.kt

```diff
@@ -0,0 +1,82 @@
+package org.wordpress.android.ui.mysite.cards.dashboard
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.dashboard.CardsStore
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.ui.mysite.MySiteSource.MySiteRefreshSource
+import org.wordpress.android.ui.mysite.MySiteUiState.PartialState.CardsUpdate
+import org.wordpress.android.ui.mysite.SelectedSiteRepository
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+class CardsSource @Inject constructor(
+    private val selectedSiteRepository: SelectedSiteRepository,
+    private val cardsStore: CardsStore,
+    @param:Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : MySiteRefreshSource<CardsUpdate> {
+    override val refresh: MutableLiveData<Boolean> = MutableLiveData<Boolean>(false)
+
+    override fun build(coroutineScope: CoroutineScope, siteLocalId: Int): LiveData<CardsUpdate> {
+        val result = MediatorLiveData<CardsUpdate>()
+        result.getData(coroutineScope, siteLocalId)
+        result.refreshData(coroutineScope, siteLocalId)
+        result.addSource(refresh) { result.refreshData(coroutineScope, siteLocalId, refresh.value) }
+        return result
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.getData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int
+    ) {
+        val selectedSite = selectedSiteRepository.getSelectedSite()
+        if (selectedSite != null && selectedSite.id == siteLocalId) {
+            coroutineScope.launch(bgDispatcher) {
+                cardsStore.getCards(selectedSite).collect { result ->
+                    postState(CardsUpdate(result))
+                }
+            }
+        }
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.refreshData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int,
+        isRefresh: Boolean? = null
+    ) {
+        when (isRefresh) {
+            null, true -> refreshData(coroutineScope, siteLocalId)
+            else -> Unit // Do nothing
+        }
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.refreshData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int
+    ) {
+        val selectedSite = selectedSiteRepository.getSelectedSite()
+        if (selectedSite != null && selectedSite.id == siteLocalId) {
```

Comment:
@ParaskP7  At this point we could have flipped the refresh flag on, how are we flipping the flag off when the selectedSite.id != siteLocalId ?

Reply:
👋 @zwarm !

Good call, although this is just us guarding from an invalid state that we will probably never reach, you are right, we should be doing that, just like it happens on `ScanAndBackupSource`. I will update the solution to add a generic error when that happens, on both places, just like we do for the try/catch on `CardsStore`:

```
if (...) {
    ...
} else {
    postState(CardsUpdate(CardsResult(CardsError(GENERIC_ERROR))))
}
```

Reply:
Added here: a5d340b28feefa89209dc361c47cb49b84cc1dd5

Reply:
🙇  It's so tricky to manage the refreshing indicator, thanks for adding this.